CHANGELOG.md

```diff
@@ -1,3 +1,8 @@
+# 0.7.0 - Bump plutus version to 1.30.0.0
+
+- Renamed `ValidatorRole` and `MintingPolicyRole` to `ThreeArgumentScript` and `TwoArgumentScript`.
```

Comment:
I don't see the point behind this change. I think it's not only unnecessary, but perhaps also backwards.

The roles communicate to the user the _purpose_ of a script. We have not removed any ambiguity, but added more. `TwoArgumentScript` does not tell us whether it's a minting policy or a staking validator, yet their usage is vastly different.

Unifying should not be done at a type level. After all, `type` aliases are almost always inferior to `newtype`s. This is the same concept. Unifying should instead be done at the instance and usage level. You could have `parseStakeValidator` and `parseMintingPolicy` both use common helpers, but it's important for both of them to actually check the role.

The argument that "runMintingPolicy" and "runStakingValidator" have the same implementation seems moot. Just alias them. They are usage site, it's fine to alias them there. Or even have `runScriptWithDatum` and `runScript` depending on how you want the interface to be. Notice how all these changes are _losing information downstream_. It doesn't seem correct to lose the information at the source.

Reply:
Furthermore, the ecosystem is much more aware of the terms "StakingValidator", "Validator", "MintingPolicy". Introducing new terms (though they are very simple terms) seems weird. I have never seen any project so far refer to them as "ThreeArgumentScript" and "TwoArgumentScript" - but this point is far less relevant than the fact that we are losing information by this transformation.

Reply:
Hi @TotallyNotChase, I'm the one who introduced the naming in the last PR. This comes from the idea that we cannot distinguish between a stake validator and a minting policy. I don't know what is more confusing: having a staking validator with the role `MintingPolicyRole` or one with the role `TwoArgumentScript`. The only information that got lost in the script role is, as far as I remember, the script's versioning. I agree that the script role does not make sense anymore. Perhaps we should drop it.

Reply:
My point is that there can be a new role for StakeValidator. Again, this is not redundant because this is essentially a _typing_ component. Same way as `newtype`s often carry invariants, therefore not being redundant.

Reply:
Why would we even bother? Sometimes, we have a script that doesn't validate a specific purpose. Depending on how you use it within a transaction, it could be used as a minting policy, staking validator, or spending validator. At this point, why do we even try to distinguish a script by its validation purpose? What if we validate two purposes, and how do we call such a script?

Reply:
Is that role description really relevant? In V3, we will have even more unified scripts. From that perspective, I don't think keeping those roles will be very useful in the future; it might increase confusion. Adding a new role for `StakeValidator` does not add much value and is not particularly future-proof IMHO

Reply:
If PlutusV3 makes the distinctions obsolete, then their type envelopes may as well not have a role, that's fair. But the v1 and v2 still need their roles (at least for a certain amount of time with support). It'd be backwards compatible anyway.

As it currently stands, most (if not all) smart contract frameworks _do_ make the distinction using types such as `Validator`, `MintingPolicy`, and `StakeValidator`. Ply's design goal has always been to be a common typing system between Haskell onchain and other transaction building offchain platforms. 

Until [some](qhttps://github.com/geniusyield/atlas/blob/3d8e6a05794c78867b216730daa7face1d4d8208/src/GeniusYield/Types/Script.hs#L9), of the [ecosystem](https://github.com/IntersectMBO/plutus-apps/tree/dbafa0ffdc1babcf8e9143ca5a7adde78d021a9a/plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts), including [plutarch itself](https://github.com/Plutonomicon/plutarch-plutus/blob/780d350f1985e89e3294861118f721d4141b2a6a/Plutarch/Api/V1.hs#L123-L129) switches to a different system, Ply has to conform. The design source is elsewhere. And because the design is consistent across the ecosystem, it wouldn't be confusing. At least, not yet.

To be clear, though I am against the idea of `TwoArgumentScript` and `ThreeArgumentScript` (due to reasons above and also because nothing in the ecosystem agrees with it), I am very much okay with removing the roles entirely. I can see it happening in the future, just _not now_, not in this PR. We have to wait for the aforementioned changes.

Could you link me to the CIP about more unified scripts please?

Reply:
Here is the CIP: [CIP-0069](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0069/)

Reply:
Ah I see, It's this one! It's been a long time but I remember this from back when Maks actually proposed it (he used to work at MLabs).

Anyway, this isn't actually an approved CIP and it has been in talks for awhile. There's no actual indication that it'll be approved and implemented. I remember Las liked the idea back when me and him were initially building Plutarch but it never ended up affecting any development since it was never even close to being actually approved.

Reply:
Unless we're talking about this one: https://github.com/cardano-foundation/CIPs/pull/784 ?

It seems like this one's different from the one linked at `cardano.org`, but it seems more promising.

Reply:
> It seems like this one's different from the one linked at cardano.org, but it seems more promising.

What do you mean? Do you see differences between the link I've shared and that PR? To me it seems the diffs into that branch were ported into cardano.org (actually I assume that page is automatically generated). Am I missing something?

Reply:
The linked CIP PR in the cardano.org page is the original old PR (the latest comment on it just says that it is not approved). But looks like the diffs are indeed from an updated PR that seems more promising.

I can see that happening in the future. And so, as I mentioned before, I'm totally open to removing roles entirely. But please not in this PR. In a separate PR in the future, when that CIP is actually implemented and the ecosystem has started moving towards it.